x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalLogicalPlanOptimizerTests.java

```diff
@@ -1073,6 +1093,146 @@ public void testPruneLeftJoinOnNullMatchingFieldAndShadowingAttributes() {
         var source = as(eval.child(), EsRelation.class);
     }
 
+    /**
+     * Expected:
+     * EsqlProject[[!alias_integer, boolean{f}#7, byte{f}#8, constant_keyword-foo{f}#9, date{f}#10, date_nanos{f}#11, dense_vector
+     * {f}#26, double{f}#12, float{f}#13, half_float{f}#14, integer{f}#16, ip{f}#17, keyword{f}#18, long{f}#19, scaled_float{f}#15,
+     * semantic_text{f}#25, short{f}#21, text{f}#22, unsigned_long{f}#20, version{f}#23, wildcard{f}#24, s{r}#5]]
+     * \_Eval[[$$dense_vector$DOTPRODUCT$27{f}#27 AS s#5]]
+     *   \_Limit[1000[INTEGER],false,false]
+     *     \_EsRelation[test_all][$$dense_vector$DOTPRODUCT$27{f}#27, !alias_integer,
+     */
+    public void testVectorFunctionsReplaced() {
+        assumeTrue("requires similarity functions", EsqlCapabilities.Cap.VECTOR_SIMILARITY_FUNCTIONS_PUSHDOWN.isEnabled());
+        String query = """
+            from test_all
+            | eval s = v_dot_product(dense_vector, [1.0, 2.0, 3.0])
```

Comment:
Suggestions, we could:
- randomize if the literal is on the left or right.
- have several evals with different vectors. (Equality of the FunctionEsField may play a role.)
- have a test case where the similarity function has to be computed on the coordinator (it shouldn't be fused then!)
- add a test case where the similarity function is nested more deeply inside the expression.
- add a test case where the exact same similarity function (including the vector literal) is used multiple times, to see if we properly deduplicate the attributes - and enhance the test case to have the vector literal once on the left and once on the right.